File: cpp/src/tutorial0_set_device_info.cpp
#include <exception>
#include <iostream>
#include "arv.h"

#define GAIN_KEY "Gain"
#define EXPOSURETIME_KEY "ExposureTime"

int main(int argc, char *argv[])
{
    arv_update_device_list ();
    unsigned int n_devices = arv_get_n_devices ();
    if (n_devices < 1){
      throw std::runtime_error("Device is not connected.");
    }

    GError *error = nullptr;

    for (unsigned int i = 0; i < n_devices; ++i){
        const char* dev_id = arv_get_device_id (i);
        ArvDevice* device = arv_open_device(dev_id, nullptr);

        printf("=== device {%d} information ===========================\n", i);

        // Get and print the current device info

        printf("%20s : %s\n",
            "Device Model Name",
            arv_device_get_string_feature_value(device, "DeviceModelName", &error));
        if (error){
            throw std::runtime_error(error->message);
        }

        double current_gain = arv_device_get_float_feature_value(device, GAIN_KEY, &error);
        if (error){
            throw std::runtime_error(error->message);
        }
        printf("%20s : %lf\n", GAIN_KEY, current_gain);
        
        double current_exposuretime = arv_device_get_float_feature_value(device, EXPOSURETIME_KEY, &error);
        if (error){
            throw std::runtime_error(error->message);
        }
        printf("%20s : %lf\n", EXPOSURETIME_KEY, current_exposuretime);
            
        printf("=>\n");

        // Set the new values of Gain and ExposureTime to the device
        double new_gain = current_gain + 10.0;
        arv_device_set_float_feature_value(device, GAIN_KEY, new_gain, &error);
        if (error){
            throw std::runtime_error(error->message);
        }

        double new_exposuretime = current_exposuretime + 20.0;
        arv_device_set_float_feature_value(device, EXPOSURETIME_KEY, new_exposuretime, &error);
        if (error){
            throw std::runtime_error(error->message);
        }

        // Get and print the current device info again
        current_gain = arv_device_get_float_feature_value(device, GAIN_KEY, &error);
        if (error){
            throw std::runtime_error(error->message);
        }
        printf("%20s : %lf\n", GAIN_KEY, current_gain);
        
        current_exposuretime = arv_device_get_float_feature_value(device, EXPOSURETIME_KEY, &error);
        if (error){
            throw std::runtime_error(error->message);
        }
        printf("%20s : %lf\n", EXPOSURETIME_KEY, current_exposuretime);
     

        g_object_unref (device);
    }

    return 0;
}

File: python/tutorial0_set_device_info.py
import os
os.add_dll_directory(os.path.join(os.environ["SENSING_DEV_ROOT"], "bin"))

import gi
gi.require_version("Aravis", "0.8")
from gi.repository import Aravis

gain_key = "Gain"
exposuretime_key = "ExposureTime"

if __name__ == "__main__":
    
    # Aravis list up the connected devices
    Aravis.update_device_list()

    num_device = Aravis.get_n_devices()
    if  num_device == 0:
        raise Exception("No device was found.")

    for i in range(num_device):
    # Access to the ith device
        device = Aravis.Camera.new(Aravis.get_device_id(i)).get_device()
        # Get GenICam feature values
        devicemodelname = device.get_string_feature_value("DeviceModelName")
        current_gain = device.get_float_feature_value(gain_key)
        current_exposuretime = device.get_float_feature_value(exposuretime_key)

        print("=== device {} information ===========================".format(i))
        print("{0:20s} : {1}".format("gain_key", current_gain))
        print("{0:20s} : {1}".format("exposuretime_key", current_exposuretime))

        print("=>")

        # Get GenICam feature values
        new_gain = current_gain + 10.0
        new_exposuretime = current_exposuretime + 20.0
        device.set_float_feature_value(gain_key, new_gain)
        device.set_float_feature_value(exposuretime_key, new_exposuretime)

        # Get GenICam feature values again
        current_gain = device.get_float_feature_value(gain_key)
        current_exposuretime = device.get_float_feature_value(exposuretime_key)

        print("{0:20s} : {1}".format("gain_key", current_gain))
        print("{0:20s} : {1}".format("exposuretime_key", current_exposuretime))

        # Release the device and close the device
        del device

    Aravis.shutdown()

